app/src/main/java/com/amaze/filemanager/asynchronous/services/ProgressiveServiceAbstract.java

```diff
@@ -0,0 +1,277 @@
+package com.amaze.filemanager.asynchronous.services;
+
+import android.app.NotificationManager;
+import android.app.PendingIntent;
+import android.app.Service;
+import android.content.Context;
+import android.content.Intent;
+import android.support.v4.app.NotificationCompat;
+import android.text.format.Formatter;
+import android.util.Log;
+
+import com.amaze.filemanager.R;
+import com.amaze.filemanager.activities.MainActivity;
+import com.amaze.filemanager.filesystem.HybridFile;
+import com.amaze.filemanager.fragments.ProcessViewerFragment;
+import com.amaze.filemanager.ui.notifications.NotificationConstants;
+import com.amaze.filemanager.utils.DatapointParcelable;
+import com.amaze.filemanager.utils.ProgressHandler;
+import com.amaze.filemanager.utils.ServiceWatcherUtil;
+
+import java.util.ArrayList;
+
+/**
+ * @author Emmanuel Messulam <emmanuelbendavid@gmail.com>
+ *         on 28/11/2017, at 19:32.
+ */
+
+public abstract class ProgressiveServiceAbstract extends Service implements ServiceWatcherUtil.ServiceWatcherInteractionInterface {
+
+    // list of data packages which contains progress
+    public NotificationManager mNotifyManager;
+    public NotificationCompat.Builder mBuilder;
+    public int notificationID;
+    public volatile float progressPercent = 0f;
+    public ProgressHandler progressHandler;
+    public Context context;
+    // list of data packages, to initiate chart in process viewer fragment
+    public ArrayList<DatapointParcelable> dataPackages;
+    public ProgressListener progressListener;
+
+    private boolean isNotificationTitleSet = false;
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+
+        initVariables();
+        return super.onStartCommand(intent, flags, startId);
+    }
+
+    public abstract void initVariables();
```

Comment:
I don't think this is a good idea... Subclasses are initializing the superclass' fields, the superclass then uses the fields blindly, it seems to me that there has to be a better way. @TranceLove ?

Reply:
Yes, I was directly using superclasses' field in subclass by declaring them public and you said opposite to this.

Reply:
Missing an initialization is not breaking enough in both, (it should lead to compile error or instant crash).

Reply:
> I don't think this is a good idea... Subclasses are initializing the superclass' fields, the superclass then uses the fields blindly, it seems to me that there has to be a better way. @TranceLove ?

I tend to use the DRY approach, as long as __I know what I'm doing__ and I know I have initialized the members properly (whenever necessary). Surely it risks difficulty in tracing when code has bugs, but subclassing with declaring fields on our own would also lead to the same result. Also, when other people wants to extend the class, they'd have to walk through a long trail figuring out why the subclass didn't behave as they'd expected.

(OK, with documentation this is not a big problem, but I'd say it's still a bad idea.)

Reply:
The other solution that comes to mind is a lot of abstract getters in superclass, or one or more different objects that can be obtained by the superclass (via abstract getters) and maintain a series of variables.

Reply:
Will try to work this around.

Reply:
Will PR tomorrow (check the branch [here](https://github.com/EmmanuelMess/AmazeFileManager/tree/fix-vars)) with a solution, it is not the best, but it's late and just fixed a bug where AS would load apk to device.